Addressed issues from QA scan report

thn-w opened the pull request.

thn-w committed:
addressed the first two categories from QA scan

thn-w committed:
addressed the rest of QA scan reports

pujav65 commented on web/web.rya/src/main/java/mvm/cloud/rdf/web/sail/RdfController.java:
```diff
@@ -266,6 +266,8 @@ private void performUpdate(String query, RepositoryConnection conn, ServletOutpu
         } catch (UpdateExecutionException e) {
             os.print(String.format("Update could not be successfully completed for query: %s\n\n", query));
+            
+            // TODO: suggest to write this to a log, not to the browser
```
sounds good -- please implement that

ghost commented:
Didn't see this addressed:
sparqlQuery.jsp, line 32 (Open Redirect)
Fortify Priority: Critical Folder Critical
Kingdom: Input Validation and Representation
Abstract: The file sparqlQuery.jsp passes unvalidated data to an HTTP redirect function on line
32. Allowing unvalidated input to control the URL used in a redirect can aid phishing
attacks.
Source: sparqlQuery.jsp:23 javax.servlet.ServletRequest.getParameter()
21 <%@ page import="java.net.*" %>
22 <%
23 String sparql=request.getParameter("sparql");
24 String infer=request.getParameter("infer");
25 String auth=request.getParameter("auth");
Sink: sparqlQuery.jsp:32 javax.servlet.http.HttpServletResponse.sendRedirect()
30 String sparqlEnc = URLEncoder.encode(sparql,"UTF-8");
31 String urlTo =
"queryrdf?query.infer="+infer+"&query.auth="+auth+"&query.resultformat="+resultFormat+
"&padding="+padding+"&query="+sparqlEnc;
32 response.sendRedirect(urlTo);
33 }
34 %>
sparqlQuery.jsp, line 32 (Open Redirect)
Fortify Priority: Critical Folder Critical
Kingdom: Input Validation and Representation
Abstract: The file sparqlQuery.jsp passes unvalidated data to an HTTP redirect function on line
32. Allowing unvalidated input to control the URL used in a redirect can aid phishing
attacks.
Source: sparqlQuery.jsp:25 javax.servlet.ServletRequest.getParameter()
23 String sparql=request.getParameter("sparql");
24 String infer=request.getParameter("infer");
25 String auth=request.getParameter("auth");
26 String resultFormat = request.getParameter("emit");
27 String padding = request.getParameter("padding");
Sink: sparqlQuery.jsp:32 javax.servlet.http.HttpServletResponse.sendRedirect()
30 String sparqlEnc = URLEncoder.encode(sparql,"UTF-8");
31 String urlTo =
"queryrdf?query.infer="+infer+"&query.auth="+auth+"&query.resultformat="+resultFormat+
"&padding="+padding+"&query="+sparqlEnc;


RyaAccumuloSailConfig.java, line 55 (Password Management: Hardcoded Password)
Fortify Priority: Critical Folder Critical
Kingdom: Security Features
Abstract: Hardcoded passwords can compromise system security in a way that cannot be easily
remedied.
Sink: RyaAccumuloSailConfig.java:55 FieldAccess: password()
53
54 private String user = "root";
55 private String password =******
56 private String instance = "dev";
57 private String zookeepers = "zoo1,zoo2,zoo3";

I'm also not sure about what was implemented for these:
NamespaceManager.java, line 59 (Unreleased Resource: Streams)
Fortify Priority: High Folder High
Kingdom: Code Quality
Abstract: The function initialize() in NamespaceManager.java sometimes fails to release a
system resource allocated by getResourceAsStream() on line 59.
Sink: NamespaceManager.java:59 cacheConfigStream =
getResourceAsStream(...)
57 this.namespaceManager = ryaDAO.getNamespaceManager();
58
59 InputStream cacheConfigStream =
Thread.currentThread().getContextClassLoader().getResourceAsStream("ehcache.xml");
60 if (cacheConfigStream == null) {
61 this.cacheManager = CacheManager.create()
RyaConsoleCommands.java, line 162 (Unreleased Resource: Streams)
Fortify Priority: High Folder High
Kingdom: Code Quality
Abstract: The function load() in RyaConsoleCommands.java sometimes fails to release a
system resource allocated by FileInputStream() on line 162.
Sink: RyaConsoleCommands.java:162 new FileInputStream(...)
160 //diff types of urls
161 try {
162 ntrips_parser.parse(new FileInputStream(file), "");
163 } catch (Exception e) {
164 LOG.log(Level.SEVERE, "", e);

And the same one on the SparqlQueryPigEngine -- I think you need to close the stream in a finally block.  

Didn't see this addressed:
SparqlQueryPigEngine.java, line 153 (Log Forging)
Fortify Priority: High Folder High
Kingdom: Input Validation and Representation
Abstract: The method runQuery() in SparqlQueryPigEngine.java writes unvalidated user input
to the log on line 153. An attacker could take advantage of this behavior to forge log
entries or inject malicious content into the log.
Source: SparqlQueryPigEngine.java:192 main(0)
190
191
192 public static void main(String[] args) {
193 try {
194 Preconditions.checkArgument(args.length == 7, "Usage: java -cp
<jar>:$PIG_LIB <class> sparqlFile hdfsSaveLocation cbinstance cbzk cbuser cbpassword
rdfTablePrefix.\n " +
Sink: SparqlQueryPigEngine.java:153 org.apache.commons.logging.Log.info()
151 Preconditions.checkNotNull(sparql, "Sparql query cannot be null");
152 Preconditions.checkNotNull(hdfsSaveLocation, "Hdfs save location cannot be
null");
153 logger.info("Running query[" + sparql + "]\n to Location[" + hdfsSaveLocation
+ "]");
154 pigServer.deleteFile(hdfsSaveLocation);
155 try {
ConfigUtils.java, line 153 (Log Forging)
Fortify Priority: High Folder High
Kingdom: Input Validation and Representation
Abstract: The method createTableIfNotExists() in ConfigUtils.java writes unvalidated user
input to the log on line 153. An attacker could take advantage of this behavior to
forge log entries or inject malicious content into the log.
Source: SparqlQueryPigEngine.java:192 main(0)
190
191
192 public static void main(String[] args) {
193 try {
194 Preconditions.checkArgument(args.length == 7, "Usage: java -cp
<jar>:$PIG_LIB <class> sparqlFile hdfsSaveLocation cbinstance cbzk cbuser cbpassword
rdfTablePrefix.\n " +
Sink: ConfigUtils.java:153 org.apache.log4j.Category.info()
151 TableOperations tops = getConnector(conf).tableOperations();
152 if (!tops.exists(tablename)) {
153 logger.info("Creating table: " + tablename);
154 tops.create(tablename);
155 return true;

You can just delete those log calls if thats easier.

Didn't see this addressed:
SparqlQueryPigEngine.java, line 197 (Path Manipulation)
Fortify Priority: High Folder High
Kingdom: Input Validation and Representation
Abstract: Attackers can control the filesystem path argument to FileInputStream() at
SparqlQueryPigEngine.java line 197, which allows them to access or modify
otherwise protected files.
Source: SparqlQueryPigEngine.java:192 main(0)
190
191
192 public static void main(String[] args) {
193 try {
194 Preconditions.checkArgument(args.length == 7, "Usage: java -cp
<jar>:$PIG_LIB <class> sparqlFile hdfsSaveLocation cbinstance cbzk cbuser cbpassword
rdfTablePrefix.\n " +
Sink: SparqlQueryPigEngine.java:197 java.io.FileInputStream.FileInputStream()
195 "Sample command: java -cp java -cp cloudbase.pig-2.0.0-SNAPSHOTshaded.
jar:/usr/local/hadoop-etc/hadoop-0.20.2/hadoop-0.20.2-core.jar:/srv_old/hdfstmp/
pig/pig-0.9.2/pig-0.9.2.jar:$HADOOP_HOME/conf
mvm.rya.accumulo.pig.SparqlQueryPigEngine " +
196 "tstSpqrl.query temp/engineTest stratus stratus13:2181 root
password l_");
197 String sparql = new String(ByteStreams.toByteArray(new
FileInputStream(args[0])));
198 String hdfsSaveLocation = args[1];
199 SparqlToPigTransformVisitor visitor = new SparqlToPigTransformVisitor();
IndexWritingTool.java, line 93 (Path Manipulation)
Fortify Priority: Critical Folder Critical
Kingdom: Input Validation and Representation
Abstract: Attackers can control the filesystem path argument to File() at IndexWritingTool.java
line 93, which allows them to access or modify otherwise protected files.
Source: IndexWritingTool.java:81 run(0)
79
80 @Override
81 public int run(final String[] args) throws Exception {
82 Preconditions.checkArgument(args.length == 7, "java " +
IndexWritingTool.class.getCanonicalName()
83 + " hdfsSaveLocation sparqlFile cbinstance cbzk cbuser cbpassword
rdfTablePrefix.");
Sink: IndexWritingTool.java:93 java.io.File.File()
91 final String tablePrefix = args[6];

Didn't see this addressed:
AccumuloStorage.java, line 91 (Password Management: Empty Password)
Fortify Priority: High Folder High
Kingdom: Security Features
Abstract: Empty passwords can compromise system security in a way that cannot be easily
remedied.
Sink: AccumuloStorage.java:91 FieldAccess: password()
89 protected String zookeepers;
90 protected String user = "";
91 protected String password =******
92 protected String table;
93 protected Text tableName;
Fortify Security Report
Copyright

Didn't see this addressed:

RdfController.java, line 269 (System Information Leak: External)
Fortify Priority: High Folder High
Kingdom: Encapsulation
Abstract: The function performUpdate() in RdfController.java might reveal system data or
debugging information by calling print() on line 269. The information revealed by
print() could help an adversary form a plan of attack.
Source: RdfController.java:269 java.lang.Throwable.getLocalizedMessage()
267 } catch (UpdateExecutionException e) {
268 os.print(String.format("Update could not be successfully completed for
query: %s\n\n", query));
269 os.print(String.format("\n\n%s", e.getLocalizedMessage()));
270 }
Sink: RdfController.java:269 javax.servlet.ServletOutputStream.print()
267 } catch (UpdateExecutionException e) {
268 os.print(String.format("Update could not be successfully completed for
query: %s\n\n", query));
269 os.print(String.format("\n\n%s", e.getLocalizedMessage()));
270 }
You can delete the logging in this case.

thn-w commented:
WRT **sparqlQuery.jsp, line 32 (Open Redirect)** - it's either we need to validate incoming data on the client side (which required format or something to indicate the data is valid or not) or on the server side. 

- If you want to validate the data on the client side, I need to know the logic to do so
- Otherwise do nothing here and let the server side validates the data

thn-w commented:
WRT **RyaAccumuloSailConfig.java, line 55 (Password Management: Hardcoded Password)** it's either it needs to be passed in or encrypted, don't know what the tool wants to do here

thn-w commented:
WRT **AccumuloStorage.java, line 91 (Password Management: Empty Password)**, here is my note in the attached file that I sent in the email

```
# Category: Password Management: Empty Password (1 Issues)
- [pig/accumulo.pig] AccumuloStorage.java
  protected String password = "";
  -> for now, let's considered this as false positive but an empty string password  could be a default password
```

Otherwise, again, we either need to pass in the value or encrypted the value. I can put an encrypted value here but not sure if fortify will complain about it or not.

thn-w committed:
Updated the code to address the top two comments

thn-w commented:
Fixed -> **RdfController.java, line 269 (System Information Leak: External)** 
Fixed -> **sparqlQuery.jsp, line 32 (Open Redirect)**

thn-w commented:
WRT **NamespaceManager.java, line 59 (Unreleased Resource: Streams)**, use try-with-resource to close the open stream when exception is thrown

thn-w committed:
Updated the code to address the rest of the issues

thn-w commented:
Fixed -> **SparqlQueryPigEngine.java, line 153 (Log Forging)**
Fixed -> **SparqlQueryPigEngine.java, line 197 (Path Manipulation)**
Fixed -> **IndexWritingTool.java, line 93 (Path Manipulation)**
Fixed -> **AccumuloStorage.java, line 91 (Password Management: Empty Password)**
